compiler/frontend/src/hir.rs

```diff
@@ -133,71 +134,125 @@ impl Body {
 #[derive(Clone, Eq, Hash, Ord, PartialEq, PartialOrd)]
 pub struct Id {
     pub module: Module,
-    pub keys: Vec<IdKey>,
+    pub keys: IdPath,
 }
+#[derive(Clone, Eq, From, Hash, Ord, PartialEq, PartialOrd, Debug)]
+pub struct IdPath(String);
 #[derive(Clone, Eq, From, Hash, Ord, PartialEq, PartialOrd)]
 pub enum IdKey {
     Named { name: String, disambiguator: usize },
     Positional(usize),
 }
+impl IdPath {
+    const fn empty() -> Self {
+        Self(String::new())
+    }
+
+    fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    fn push(&self, key: IdKey) -> Self {
```

Comment:
`push(…)` sounds like it's modifying this instance. Maybe rename it to `child(…)`? And please add `#[must_use]`